### PR TITLE
HPCC-16226 WsDESDLConfig methods should echo resulting binding/def

### PR DESCRIPTION
**Pull request by rpastrana**
- Provides option to echo binding upon publish, or configuration change
- Provides option to echo defintion upon publish

Signed-off-by: rpastrana rodrigo.pastrana@lexisnexis.com

### CODE REVIEW

File: esp/scm/ws_esdlconfig.ecm [hunk 1]
```diff
@@ -25,20 +25,42 @@ ESPStruct MethodConfig
...
     String Name;
     ESParray<ESPstruct NamedValue, Attribute> Attributes;
+    String Elements;
 };
 
 ESPStruct ESDLConfiguration
 {
     ESParray<ESPStruct MethodConfig, Method> Methods;
 };
+ESPstruct ESDLBinding
```
Comment (rpastrana) [inline esp/scm/ws_esdlconfig.ecm]: pre-existing, moved up in file for visibility

File: esp/scm/ws_esdlconfig.ecm [hunk 2]
```diff
@@ -25,20 +25,42 @@ ESPStruct MethodConfig
...
+    string EspBinding;
+    string Id;
+};
+
+ESPstruct ESDLDefinition
+{
+    String Name;
+    int Seq;
+    String Id;
+    [min_ver("1.1")] String Interface;
```
Comment (rpastrana) [inline esp/scm/ws_esdlconfig.ecm]: pre-existing, moved up in file for visibility

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 3]
```diff
@@ -217,6 +217,32 @@ bool ensureSDSPath(const char * sdsPath)
     return true;
 }
 
+void fetchESDLDefinitionFromDaliById(const char *id, StringBuffer & def)
```
Comment (rpastrana) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: pre-existing, moved up in file for visibility

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 4]
```diff
@@ -476,20 +502,102 @@ bool CWsESDLConfigEx::onPublishESDLDefinition(IEspContext &context, IEspPublishE
...
+            {
+                msg.set("Could not publish ESDL Definition, unable to fetch ESDL Definition registry.");
+                resp.updateStatus().setCode(-1);
+            }
         }
-        else
+
+        msg.appendf("Successfully published %s", newqueryid.str());
+
+        double ver = context.getClientVersion();
```
Comment (rpastrana) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: after publishing the definition, fetch it from Dali and report it in response

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 5]
```diff
@@ -955,6 +1170,112 @@ bool CWsESDLConfigEx::onConfigureESDLBindingMethod(IEspContext &context, IEspCon
                     StringBuffer msg;
                     publishESDLMethod(espProcName.str(), espBindingName.str(), esdlDefIdSTR.toLowerCase().str(), methodName, &item, override, msg);
                     resp.updateStatus().setDescription(msg.str());
+
+                    double ver = context.getClientVersion();
```
Comment (rpastrana) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: after configuring a method within a binding (ESDLBinding/Definition/Methods/Method) fetch the entire binding, and report it in the response

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 6]
```diff
@@ -1040,32 +1361,6 @@ int CWsESDLConfigEx::getBindingXML(const char * espProcName, const char * espBin
     }
 }
 
-void fetchESDLDefinitionFromDaliById(const char *id, StringBuffer & def)
```
Comment (rpastrana) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: moved up for visibility

**Comment by rpastrana** [comment]
@wangkx please review.

File: esp/scm/ws_esdlconfig.ecm [hunk 7]
```diff
@@ -25,20 +25,42 @@ ESPStruct MethodConfig
 {
     String Name;
     ESParray<ESPstruct NamedValue, Attribute> Attributes;
+    String Elements;
```
Comment (wangkx) [inline esp/scm/ws_esdlconfig.ecm]: versioning?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 8]
```diff
@@ -476,20 +502,102 @@ bool CWsESDLConfigEx::onPublishESDLDefinition(IEspContext &context, IEspPublishE
...
+            }
+            else
+            {
+                msg.set("Could not publish ESDL Definition, unable to fetch ESDL Definition registry.");
+                resp.updateStatus().setCode(-1);
+            }
         }
-        else
+
+        msg.appendf("Successfully published %s", newqueryid.str());
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: In line 516, msg says "Could not publish...". But, "Successfully published..." appended here?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 9]
```diff
@@ -727,6 +835,113 @@ bool CWsESDLConfigEx::onPublishESDLBinding(IEspContext &context, IEspPublishESDL
...
+                if (req.getEchoBinding())
+                {
+                    StringBuffer msg;
+                    Owned<IPropertyTree> esdlbindingtree = getBindingTree(espProcName.str(), espBindingName.str(), msg);
+                    if (esdlbindingtree)
+                    {
+                        IArrayOf<IEspMethodConfig> iesmethods;
+
+                        Owned<IPropertyTree> def;
+                        def.setown(esdlbindingtree->queryPropTree("Definition[1]"));
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: queryPropTree() should not setown?

File: esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp [hunk 10]
```diff
@@ -955,6 +1170,112 @@ bool CWsESDLConfigEx::onConfigureESDLBindingMethod(IEspContext &context, IEspCon
...
+                    if (ver >= 1.2)
+                    {
+                        StringBuffer msg;
+                        Owned<IPropertyTree> esdlbindingtree = getBindingTree(espProcName.str(), espBindingName.str(), msg);
+                        if (esdlbindingtree)
+                        {
+                            IArrayOf<IEspMethodConfig> iesmethods;
+
+                            Owned<IPropertyTree> def;
+                            def.setown(esdlbindingtree->queryPropTree("Definition[1]"));
```
Comment (wangkx) [inline esp/services/ws_esdlconfig/ws_esdlconfigservice.cpp]: Same as line 849

**Comment by wangkx** [comment]
@rpastrana looks good. A few comments.

File: esp/scm/ws_esdlconfig.ecm (hunk 7 above)
Comment (rpastrana) [inline esp/scm/ws_esdlconfig.ecm]: MethodConfig was added in 1.2, and all uses of the structure are versioned, so I think we're fine.

**Comment by rpastrana** [comment]
@wangkx please review code review changes.

**Comment by wangkx** [comment]
@rpastrana the changes look fine.

**Comment by richardkchapman** [comment]
Again, commit title could be clearer

**Comment by rpastrana** [comment]
@richardkchapman title updated, please merge.

**Pull request merged**